tests/system_tests_multi_tenancy.py

```diff
@@ -229,6 +231,17 @@ def test_16_two_router_anonymous_sender_tenant_on_both(self):
         test.run()
         self.assertIsNone(test.error)
 
+    def _cleanup_link_routes(func):
```

Comment:
The 'traditional' way of using `unittest` would be to do your `setUp()` and `tearDown()` in these two methods. (With classes being understood as a way to group together tests that share setup and teardown.)

I'd probably try to put the teardown into `LinkRouteTest` itself... but that does not have access to `self.routers`, so it is not such a good idea. Maybe add method `def tearDown(routers)` to it and call it in every test...

Decorators are too much magic for my simple tastes, I guess.

It's a matter of preference. The decorator is not too hard to understand... And if I had unlimited time, enthusiasm, and money to redesign this test "the best way", I would not actually know what to do with it right away... So, lgtm, just go ahead, I guess?

Reply:
> Yeah in the end I'm punting on the decorator approach: too many issues with the python linters.

If the only problem is the linter, that should not be hard to put right. Put the decorator as a standalone top level method in the module, and the warning you got should be gone! You were not relying on it being an instance method anyway.

```python
def _cleanup_link_routes(func):
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        result = func(self, *args, **kwargs)
        # TODO something here
        return result
    return wrapper

class TestClass(unittest.TestCae):
    def __init__(self, a=1, b=2):
    @decorate
    def test_something(self):
        pass
```

I'm not sure what other problems there might be with it. I'd have to actually try it, and I don't want to poke into tests that I don't understand.